File: src/main/java/com/teambrmodding/assistedprogression/client/gui/storage/GuiFlushableChest.scala
package com.teambrmodding.assistedprogression.client.gui.storage

import java.awt.Color

import com.teambr.bookshelf.client.gui.component.BaseComponent
import com.teambr.bookshelf.client.gui.component.control.{GuiComponentCheckBox, GuiComponentSetNumber}
import com.teambr.bookshelf.client.gui.component.display.{GuiComponentText, GuiTabCollection}
import com.teambr.bookshelf.client.gui.{GuiBase, GuiColor}
import com.teambrmodding.assistedprogression.common.container.storage.ContainerFlushableChest
import com.teambrmodding.assistedprogression.common.tiles.storage.TileFlushableChest
import net.minecraft.client.resources.I18n
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.init.Blocks
import net.minecraft.item.{Item, ItemStack}

import scala.collection.mutable.ArrayBuffer

/**
  * This file was created for AssistedProgression
  *
  * AssistedProgression is licensed under the
  * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
  * http://creativecommons.org/licenses/by-nc-sa/4.0/
  *
  * @author Dyonovan
  * @since 1/16/2017
  */
class GuiFlushableChest(player: EntityPlayer, tile: TileFlushableChest) extends
        GuiBase[ContainerFlushableChest](new ContainerFlushableChest(player.inventory, tile), 175, 165, "assistedprogression.flushchest.title"){

    override def addComponents(): Unit = { }

    override def addRightTabs(tabs: GuiTabCollection): Unit = {
        val infoTab = new ArrayBuffer[BaseComponent]()
        infoTab += new GuiComponentSetNumber(72, 21, 30, tile.getFlushInterval, 1, 60) {
            override def setValue(i: Int): Unit = {
                tile.setFlushInterval(i)
                tile.sendValueToServer(tile.FLUSH_INTERVAL, i.toDouble)
            }
        }
        infoTab += new GuiComponentCheckBox(10, 20, "AutoFlush", tile.getAutoFlush) {
            override def setValue(bool: Boolean): Unit = {
                tile.setAutoFlush(bool)
                tile.sendValueToServer(tile.AUTO_FLUSH, if (bool) 0 else -1)
            }
        }
        infoTab += new GuiComponentCheckBox(10, 30, "Sound", tile.getFlushSound) {
            override def setValue(bool: Boolean): Unit = {
                tile.setFlushSound(bool)
                tile.sendValueToServer(tile.FLUSH_SOUND, if (bool) 0 else -1)
            }
        }
        infoTab += new GuiComponentText(GuiColor.ORANGE + I18n.format("assistedprogression.text.config"), 26, 6)
        tabs.addTab(infoTab.toList, 115, 50, new Color(0, 155,0), new ItemStack(Item.getItemFromBlock(Blocks.STONE_BUTTON)))
    }
}

File: src/main/java/com/teambrmodding/assistedprogression/client/gui/storage/GuiFlushableChest.java
package com.teambrmodding.assistedprogression.client.gui.storage;

import com.teambr.bookshelf.client.gui.GuiBase;
import com.teambr.bookshelf.client.gui.GuiColor;
import com.teambr.bookshelf.client.gui.component.BaseComponent;
import com.teambr.bookshelf.client.gui.component.control.GuiComponentCheckBox;
import com.teambr.bookshelf.client.gui.component.control.GuiComponentSetNumber;
import com.teambr.bookshelf.client.gui.component.display.GuiComponentText;
import com.teambr.bookshelf.client.gui.component.display.GuiTabCollection;
import com.teambr.bookshelf.util.ClientUtils;
import com.teambrmodding.assistedprogression.common.container.storage.ContainerFlushableChest;
import com.teambrmodding.assistedprogression.common.tiles.storage.TileFlushableChest;
import com.teambrmodding.assistedprogression.lib.Reference;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.init.Blocks;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.util.ResourceLocation;

import java.util.ArrayList;

/**
 * This file was created for AssistedProgression
 * <p>
 * AssistedProgression is licensed under the
 * Creative Commons Attribution-NonCommercial-ShareAlike 4.0 International License:
 * http://creativecommons.org/licenses/by-nc-sa/4.0/
 *
 * @author Paul Davis - pauljoda
 * @since 11/13/17
 */
public class GuiFlushableChest extends GuiBase<ContainerFlushableChest> {
    private TileFlushableChest chest;

    /**
     * Main constructor for Guis
     *
     * @param inventory The container
     * @param width     The width of the gui
     * @param height    The height of the gui
     * @param title     The title of the gui
     * @param texture   The location of the background texture
     */
    public GuiFlushableChest(EntityPlayer player, TileFlushableChest tile) {
        super(new ContainerFlushableChest(player.inventory, tile), 175, 165,
                tile.getBlockType().getLocalizedName(),
                new ResourceLocation(Reference.MOD_ID, "textures/gui/flushable_chest.png"));
        chest = tile;
    }

    /**
     * This will be called after the GUI has been initialized and should be where you add all components.
     */
    @Override
    protected void addComponents() {
        // No Op
    }

    @Override
    protected void addRightTabs(GuiTabCollection tabs) {
        ArrayList<BaseComponent> infoTabComponents = new ArrayList<>();

        // Interval
        infoTabComponents.add(new GuiComponentSetNumber(this, 72, 21, 177, 0,
                chest.getVariable(TileFlushableChest.FLUSH_INTERVAL_VARID).intValue(), 1, 60) {
            /**
             * Called when the user sets the value or when the value is changed
             *
             * @param value The value set by the user
             */
            @Override
            protected void setValue(int value) {
                chest.setVariable(TileFlushableChest.FLUSH_INTERVAL_VARID, value);
                chest.sendValueToServer(TileFlushableChest.FLUSH_INTERVAL_VARID, value);
            }
        });

        // Auto Flush
        infoTabComponents.add(new GuiComponentCheckBox(this, 10, 20, 178, 22,
                chest.getVariable(TileFlushableChest.AUTO_FLUSH_VARID) == 1 ? true : false, ClientUtils.translate("assistedprogression.text.flushableChest.autoFlush")) {
            @Override
            protected void setValue(boolean value) {
                chest.setVariable(TileFlushableChest.AUTO_FLUSH_VARID, value ? 1 : -1);
                chest.sendValueToServer(TileFlushableChest.AUTO_FLUSH_VARID, value ? 1 : -1);
            }
        });

        // Play Sounds
        infoTabComponents.add(new GuiComponentCheckBox(this, 10, 30, 178, 22,
                chest.getVariable(TileFlushableChest.PLAY_FLUSH_SOUND_VARID) == 1 ? true : false, ClientUtils.translate("assistedprogression.text.flushableChest.playSound")) {
            @Override
            protected void setValue(boolean value) {
                chest.setVariable(TileFlushableChest.PLAY_FLUSH_SOUND_VARID, value ? 1 : -1);
                chest.sendValueToServer(TileFlushableChest.PLAY_FLUSH_SOUND_VARID, value ? 1 : -1);
            }
        });

        // Title
        infoTabComponents.add(new GuiComponentText(this, 26, 6,
                GuiColor.ORANGE + ClientUtils.translate("assistedprogression.text.config"), null));
        tabs.addTab(infoTabComponents, 115, 50, 176, 37, new ItemStack(Item.getItemFromBlock(Blocks.STONE_BUTTON)));
    }
}
